File: archive/canvas-api/src/main/java/com/instructure/canvasapi/model/CanvasContext.java
/*
 * Copyright (C) 2017 - present Instructure, Inc.
 *
 *     Licensed under the Apache License, Version 2.0 (the "License");
 *     you may not use this file except in compliance with the License.
 *     You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *     Unless required by applicable law or agreed to in writing, software
 *     distributed under the License is distributed on an "AS IS" BASIS,
 *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *     See the License for the specific language governing permissions and
 *     limitations under the License.
 *
 */

package com.instructure.canvasapi.model;

import java.util.Date;

public abstract class CanvasContext extends CanvasModel<CanvasContext> implements android.os.Parcelable {

    public static final String HOME_FEED = "feed";
    public static final String HOME_WIKI = "wiki";
    public static final String HOME_MODULES = "modules";
    public static final String HOME_ASSIGNMENTS = "assignments";
    public static final String HOME_SYLLABUS = "syllabus";

    public enum Type {
        GROUP, COURSE, USER, SECTION, UNKNOWN;

        public static boolean isGroup(CanvasContext canvasContext) {
            if (canvasContext == null) {
                return false;
            }
            return canvasContext.getType() == GROUP;
        }

        public static boolean isCourse(CanvasContext canvasContext) {
            if (canvasContext == null) {
                return false;
            }
            return canvasContext.getType() == COURSE;
        }

        public static boolean isUser(CanvasContext canvasContext) {
            if (canvasContext == null) {
                return false;
            }
            return canvasContext.getType() == USER;
        }

        public static boolean isUnknown(CanvasContext canvasContext) {
            if (canvasContext == null) {
                return false;
            }
            return canvasContext.getType() == UNKNOWN;
        }

        public static boolean isSection(CanvasContext canvasContext) {
            if (canvasContext == null) {
                return false;
            }
            return canvasContext.getType() == SECTION;
        }
    }


    public abstract String getName();

    public abstract Type getType();

    public abstract long getId();

    protected String default_view;

    protected CanvasContextPermission permissions;

    public void setPermissions(CanvasContextPermission permissions) {
        this.permissions = permissions;
    }

    public CanvasContextPermission getPermissions() {
        return permissions;
    }

    public boolean canCreateDiscussion() {
        return (permissions != null && permissions.canCreateDiscussionTopic());
    }

    ///////////////////////////////////////////////////////////////////////////
    // Overrides
    ///////////////////////////////////////////////////////////////////////////

    @Override
    public Date getComparisonDate() {
        return null;
    }

    @Override
    public String getComparisonString() {
        return getName();
    }

    /**
     * Make sure they have the same type and the same ID.
     */
    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        if (!super.equals(o)) return false;

        CanvasContext that = (CanvasContext) o;

        return !(getType() != that.getType() || getId() != that.getId());

    }

    ///////////////////////////////////////////////////////////////////////////
    // Helpers
    ///////////////////////////////////////////////////////////////////////////

    /**
     * For courses, returns the course code.
     * For everything else, returns the Name;
     */

    public String getSecondaryName() {
        String secondaryName = getName();
        if (getType() == CanvasContext.Type.COURSE) {
            secondaryName = ((Course) this).getCourseCode();
        }
        return secondaryName;
    }


    /**
     * Used for Cache Filenames in the API.
     */
    public String toAPIString() {
        String typeString;
        if (getType().equals(Type.GROUP)) {
            typeString = "groups";
        } else if (getType().equals(Type.COURSE)) {
            typeString = "courses";
        } else if (getType().equals(Type.SECTION)) {
            typeString = "sections";
        } else {
            typeString = "users";
        }

        String idString = Long.toString(getId());
        if (getType() == Type.USER && getId() == 0) {
            idString = "self";
        }

        return "/" + typeString + "/" + idString;
    }

    /**
     * @returns group_:id or course_:id
     */
    public String getContextId() {

        String prefix = "";
        if (getType() == Type.COURSE) {
            prefix = "course";
        } else if (getType() == Type.GROUP) {
            prefix = "group";
        } else if(getType() == Type.USER) {
            prefix = "user";
        }

        return prefix + "_" + getId();
    }

    public static String makeContextId(Type type, long id) {
        String prefix = "";
        if (type == Type.COURSE) {
            prefix = "course";
        } else if (type == Type.GROUP) {
            prefix = "group";
        } else if(type == Type.USER) {
            prefix = "user";
        } else {
            return null;
        }

        return prefix + "_" + id;
    }

    /**
     * Get home page label returns the fragment identifier.
     *
     * @return
     */
    public String getHomePageID() {
        if (default_view == null) {
            //notifications can't be hidden, so if for some reason we don't have the home page
            //send them to notifications instead
            return Tab.NOTIFICATIONS_ID;
        }
        if (default_view.equals(HOME_FEED)) {
            return Tab.NOTIFICATIONS_ID;
        }
        if (default_view.equals(HOME_SYLLABUS)) {
            return Tab.SYLLABUS_ID;
        }
        if (default_view.equals(HOME_WIKI)) {
            return Tab.PAGES_ID;
        }
        if (default_view.equals(HOME_ASSIGNMENTS)) {
            return Tab.ASSIGNMENTS_ID;
        }
        if (default_view.equals(HOME_MODULES)) {
            return Tab.MODULES_ID;
        }
        return Tab.NOTIFICATIONS_ID; //send them to notifications if we don't know what to do
    }

    public static CanvasContext getGenericContext(final Type type, final long id, final String name) {
        CanvasContext canvasContext;
        if(type == Type.USER){
          User  user = new User(id);
            user.setName(name);
            canvasContext = user;
        } else if (type == Type.COURSE){
            Course course = new Course();
            course.setId(id);
            course.setName(name);

            canvasContext = course;
        } else if (type == Type.GROUP){
            Group group = new Group();
            group.setId(id);
            group.setName(name);

            canvasContext = group;
        } else if (type == Type.SECTION){
            Section section = new Section();
            section.setId(id);
            section.setName(name);

            canvasContext = section;
        } else {
            return null;
        }

        return canvasContext;
    }

    public static CanvasContext emptyCourseContext() {
        return getGenericContext(Type.COURSE, 0, "");
    }

    public static CanvasContext emptyUserContext() {
        return getGenericContext(Type.USER, 0, "");
    }


}

File: libs/canvas-api-2/src/main/java/com/instructure/canvasapi2/models/CanvasContext.kt
/*
 * Copyright (C) 2017 - present Instructure, Inc.
 *
 *     Licensed under the Apache License, Version 2.0 (the "License");
 *     you may not use this file except in compliance with the License.
 *     You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *     Unless required by applicable law or agreed to in writing, software
 *     distributed under the License is distributed on an "AS IS" BASIS,
 *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *     See the License for the specific language governing permissions and
 *     limitations under the License.
 *
 */

package com.instructure.canvasapi2.models

import com.instructure.canvasapi2.utils.ApiPrefs
import java.util.Date
import java.util.Locale

abstract class CanvasContext : CanvasModel<CanvasContext>() {
    abstract val name: String?
    abstract val type: Type
    abstract override val id: Long

    var permissions: CanvasContextPermission? = null

    override val comparisonDate: Date? get() = null
    override val comparisonString: String? get() = name

    /**
     * For courses, returns the course code.
     * For everything else, returns the Name;
     */

    val secondaryName: String?
        get() {
            var secondaryName: String? = name
            if (type == CanvasContext.Type.COURSE) {
                secondaryName = (this as Course).courseCode
            }
            return secondaryName
        }

    /**
     * @returns group_:id or course_:id
     */
    val contextId: String
        get() {
            val prefix: String = when (type) {
                Type.COURSE -> "course"
                Type.GROUP -> "group"
                Type.USER -> "user"
                Type.SECTION -> TODO()
                Type.UNKNOWN -> "unknown"
            }

            return prefix + "_" + id
        }

    /**
     * Make sure they have the same type and the same ID.
     */
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        if (!super.equals(other)) return false

        val that = other as CanvasContext

        return !(type != that.type || id != that.id)
    }

    enum class Type(val apiString: String) {
        GROUP("groups"),
        COURSE("courses"),
        USER("users"),
        SECTION("sections"),
        UNKNOWN("unknown");

        companion object {
            fun isGroup(canvasContext: CanvasContext?): Boolean = GROUP == canvasContext?.type
            fun isCourse(canvasContext: CanvasContext?): Boolean = COURSE == canvasContext?.type
            fun isUser(canvasContext: CanvasContext?): Boolean = USER == canvasContext?.type
            fun isUnknown(canvasContext: CanvasContext?): Boolean = UNKNOWN == canvasContext?.type
            fun isSection(canvasContext: CanvasContext?): Boolean = SECTION == canvasContext?.type
        }
    }

    fun canCreateDiscussion(): Boolean = permissions != null && permissions!!.canCreateDiscussionTopic

    /**
     * Used for Cache Filenames in the API.
     */
    fun toAPIString(): String {
        val idString = if (type == Type.USER && id == 0L) {
            "self"
        } else id.toString()
        return "/${type.apiString}/$idString"
    }

    fun apiContext(): String = if (type == Type.COURSE) "courses" else "groups"

    companion object {

        fun makeContextId(type: Type, id: Long): String = type.name.lowercase(Locale.getDefault()) + "_" + id

        /**
         * Returns a generic CanvasContext based on the provided context code.
         * @param contextCode Context code string, e.g. "course_1"
         * @return A generic CanvasContext, or null if the provided contextCode is invalid
         */
        fun fromContextCode(contextCode: String?, name: String? = ""): CanvasContext? {
            if (contextCode.isNullOrBlank() || name == null) return null

            val codeParts = contextCode.split("_".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            if (codeParts.size != 2) return null

            val type = when (codeParts[0]) {
                "course" -> Type.COURSE
                "group" -> Type.GROUP
                "user" -> Type.USER
                else -> Type.UNKNOWN
            }

            val id: Long
            try {
                id = codeParts[1].toLong()
            } catch (e: NumberFormatException) {
                return null
            }

            return getGenericContext(type, id, name)
        }

        fun getApiContext(canvasContext: CanvasContext): String = if (canvasContext.type == Type.COURSE) "courses" else "groups"

        fun getGenericContext(type: Type, id: Long = -1L, name: String = ""): CanvasContext =
                when (type) {
                    Type.USER -> User(id = id, name = name)
                    Type.COURSE -> Course(id = id, name = name)
                    Type.GROUP -> Group(id = id, name = name)
                    Type.SECTION -> Section(id = id, name = name)
                    Type.UNKNOWN -> Unknown()
                }

        @JvmOverloads
        fun emptyCourseContext(id: Long = 0L): CanvasContext = getGenericContext(Type.COURSE, id, "")
        fun emptyGroupContext(id: Long = 0L): CanvasContext = getGenericContext(Type.GROUP, id, "")
        fun defaultCanvasContext(): CanvasContext = getGenericContext(Type.UNKNOWN, 0, "")
        fun emptyUserContext(): CanvasContext = getGenericContext(Type.USER, 0, "")
        fun currentUserContext(user: User): CanvasContext = getGenericContext(Type.USER, user.id, user.name)
    }
}

/**
 * Creates a base URL for this CanvasContext using the current domain and context type/id
 */
fun CanvasContext.toBaseUrl(): String {
    return "${ApiPrefs.fullDomain}/${type.apiString}/$id"
}
